File: src/nif_allocator.c
#include "nif_allocator.h"
#include "erl_nif.h"
#include <string.h>

void* nif_alloc(size_t size) {
    return enif_alloc(size);
}

void* nif_calloc(size_t num, size_t size) {
    size_t total_size = num * size;
    void* ptr = enif_alloc(total_size);
    if (ptr) {
        memset(ptr, 0, total_size);
    }
    return ptr;
}

void* nif_realloc(void* ptr, size_t new_size) {
    return enif_realloc(ptr, new_size);
}

void nif_free(void* ptr) {
    enif_free(ptr);
}

File: src/nif_allocator.zig
const e = @import("./erl_nif.zig");

pub export fn nif_alloc(size: usize) callconv(.C) ?*anyopaque {
    return e.enif_alloc(size);
}

pub export fn nif_calloc(num: usize, size: usize) callconv(.C) ?*anyopaque {
    const total_size = num * size;
    const ptr = e.enif_alloc(total_size);
    if (ptr != null) {
        @memset(@as([*]u8, @ptrCast(ptr))[0..total_size], 0);
    }
    return ptr;
}

pub export fn nif_realloc(ptr: ?*anyopaque, new_size: usize) callconv(.C) ?*anyopaque {
    return e.enif_realloc(ptr, new_size);
}

pub export fn nif_free(ptr: ?*anyopaque) callconv(.C) void {
    e.enif_free(ptr);
}
